website/blog/2020-08-04-prestodb-and-hudi.md

```diff
@@ -0,0 +1,153 @@
+- Inserts to parquet files - This is done for tables that do not have indexable log files, for eg bloom index
+
+At a later time, the log files are merged with the base parquet file by compaction action in the timeline. This table type is the most versatile, highly advanced and offers much flexibility for writing (ability to specify different compaction policies, absorb bursty write traffic etc) and querying (e.g: tradeoff data freshness and query performance). At the same time, it can involve a learning curve for mastering it operationally. 
+
+## Early Presto integration
+
+Hudi was designed in mid to late 2016. At that time, we were looking to integrate with query engines in the Hadoop ecosystem. To achieve this in Presto, we introduced a custom annotation - `@UseFileSplitsFromInputFormat`, as suggested by the community. Any Hive registered table if it has this annotation would fetch splits by invoking the corresponding inputformat’s `getSplits()` method instead of Presto Hive’s native split loading logic. For Hudi tables queried via Presto this would be a simple call to `HoodieParquetInputFormat.getSplits()`. This was a straightforward and simple integration. All one had to do was drop in corresponding Hudi jars under `<presto_install>/plugin/hive-hadoop2/` directory. This supported querying COW Hudi tables and read optimized querying of MOR Hudi tables (only fetch data from compacted base parquet files). At Uber, this simple integration already supported over 100,000 Presto queries per day from 100s of petabytes of data (raw data and modeled tables) sitting in HDFS ingested using Hudi. 
+
+## Moving away from InputFormat.getSplits()
+
+While invoking `inputformat.getSplits()` was a simple integration, we noticed that this could cause a lot of RPC calls to namenode. There were several disadvantages to the previous approach. 
+
+1. The `InputSplit`s returned from Hudi are not enough. Presto requires to know the file status and block locations for each of the `InputSplit` returned. So this added 2 extra RPCs to the namenode for every split times the number of partitions loaded. Occasionally, backpressure can be observed if the namenode is under lots of pressure.
+2. Further, for every partition loaded (per `loadPartition()` call) in Presto split calculation, `HoodieParquetInputFormat.getSplits()` would be invoked. That caused redundant Hoodie table metadata listing, which otherwise can be reused for all partitions belonging to a table scanned from a query. 
+
+This led us to rethink the Presto-Hudi integration. At Uber, we changed this implementation by taking compile time dependency on Hudi and instantiated the `HoodieTableMetadata` once in the `BackgroundHiveSplitLoader` constructor. And leveraged Hudi Library APIs to filter the partition files instead of calling `HoodieParquetInputFormat.getSplits()`. This gave a significant reduction in the number of namenode calls in this path. 
```

Comment:
taking compile time -> adding a compile time
And leveraged -> We then leveraged